tests/vlan/test_vlan.py

```diff
@@ -35,71 +33,162 @@ def cfg_facts(duthosts, rand_one_dut_hostname):
+    config_port_channel_members = [port_channel['members'] for port_channel in config_portchannels.values()]
     config_port_channel_member_ports = list(itertools.chain.from_iterable(config_port_channel_members))
+    # key is dev name, value is list for configured VLAN member.
+    config_ports_vlan = defaultdict(list)
+    vlan_members = cfg_facts.get('VLAN_MEMBER', {})
+    for k, v in cfg_facts['VLAN'].items():
+        vlanid = v['vlanid']
+        for addr in cfg_facts['VLAN_INTERFACE']['Vlan'+vlanid]:
+            # address could be IPV6 and IPV4, only need IPV4 here
+            if addr.find(':') == -1:
+                ip = addr
+                break
+        else:
+            continue
+        for port in v['members']:
+            if k in vlan_members and port in vlan_members[k]:
+                if 'tagging_mode' not in vlan_members[k][port]:
+                    continue
+                mode = vlan_members[k][port]['tagging_mode']
+                config_ports_vlan[port].append({'vlanid':int(vlanid), 'ip':ip, 'tagging_mode':mode})
+
+    # For t0 topo, will add port to new VLAN.
+    # pvid_cycle uses vlan_id_list to iterate new VLAN id.
+    vlan_id_list = [k for k, v in vlan_intfs_dict.items() if v['orig'] == False]
     pvid_cycle = itertools.cycle(vlan_id_list)
     # when running on t0 we can use the portchannel members
     if config_portchannels:
-        for po in config_portchannels.keys()[:2]:
+        portchannel_cnt = 0
+        for po in config_portchannels:
             port = config_portchannels[po]['members'][0]
-            vlan_ports_list.append({
+            port_id = config_port_indices.keys().index(port)
+            vlan_port = {
                 'dev' : po,
                 'port_index' : [config_port_indices[member] for member in config_portchannels[po]['members']],
-                'pvid' : pvid_cycle.next(),
-                'permit_vlanid' : { vid : {
-                    'peer_ip' : '192.168.{}.{}'.format(vid, 2 + config_port_indices.keys().index(port)),
-                    'remote_ip' : '{}.1.1.{}'.format(vid, 2 + config_port_indices.keys().index(port))
-                    } for vid in vlan_id_list }
-            })
+                'permit_vlanid' : {}
+            }
+            if tbinfo['topo']['name'] == 't0-56-po2vlan' and po in config_ports_vlan:
+                vlan_port['pvid'] = 0
+                for vlan in config_ports_vlan[po]:
+                    if 'vlanid' not in vlan or 'ip' not in vlan or 'tagging_mode' not in vlan:
+                        continue
+                    if vlan['tagging_mode'] == 'untagged':
+                        vlan_port['pvid'] = vlan['vlanid']
+                    ip_prefix = '.'.join(vlan['ip'].split('.')[:3])
+                    vlan_port['permit_vlanid'][vlan['vlanid']] = {
+                            'peer_ip' : '{}.{}'.format(ip_prefix, 2 + port_id),
+                            'remote_ip' : '{}.1.1.{}'.format(vlan['vlanid']&255, 2 + port_id)
```

Comment:
>vlan['vlanid']&255 [](http://example.com/codeflow?start=61&length=18)

What happens if vlanid & 255 conflict? Is it a disaster?

Reply:
vlanid & 255 can guarantee the result is no greater than 255, just use to generate remote_ip.